File: python/pyxis/_datum_compense.py
from typing import overload

import numpy as np
from numpy.typing import NDArray

from ._utils import TCoordArray, TCoordScalar, coord_util


@overload
def datum_compense(
    xc: TCoordScalar,
    yc: TCoordScalar,
    hb: float,
    radius: float,
    x0: float,
    y0: float,
    clone: bool = True,
) -> tuple[float, float]: ...
@overload
def datum_compense(
    xc: TCoordArray,
    yc: TCoordArray,
    hb: float,
    radius: float,
    x0: float,
    y0: float,
    clone: bool = True,
) -> tuple[NDArray[np.float64], NDArray[np.float64]]: ...
def datum_compense(
    xc,
    yc,
    hb: float,
    radius: float,
    x0: float,
    y0: float,
    clone: bool = True,
):
    r"""
    Convert projected XY coordinates from height compensation plane to sea level plane.

    Unit: meter

    Parameters
    ----------
    xc, yc
        Coordinates on height compensation plane.
    hb
        Elevation of height compensation plane.
    radius
        Radius of earth.
    x0, y0
        Coordinate system origin.
    clone
        If True, a deep copy of `data` will be created before processing.
        If False, `data` will be modified in place. Default is False.

    Returns
    -------
    tuple[Any, Any]
        Projected XY coordinates of sea level plane.

    References
    ----------
    .. [1] 杨元兴.抵偿高程面的选择与计算[J].城市勘测,2008(02):72-74.

    Examples
    --------
    >>> from glatzel import pyxis
    >>> pyxis.datum_compense(469704.6693, 2821940.796, 400)
    2821763.831232311 469706.56912942487
    >>> pyxis.datum_compense(
    ...     np.array([469704.6693, 469704.6693]),
    ...     np.array([2821940.796, 2821940.796]),
    ...     400,
    ... )
    ( array([469706.56912942, 469706.56912942]), array([2821763.83123231, 2821763.83123231]),)
    """
    from .pyxis_py import py_datum_compense  # type: ignore

    xc = coord_util("xc", xc, clone)
    yc = coord_util("yc", yc, clone)

    xc, yc = py_datum_compense(xc, yc, hb, radius, x0, y0)

    return xc, yc

File: rust/crates/pyxis/src/datum_compense.rs
use crate::GeoFloat;
pub trait IDatumCompenseParms<T: GeoFloat> {
    fn x0(&self) -> T;
    fn y0(&self) -> T;
    fn factor(&self) -> T;
}
pub struct DatumCompenseParms<T: GeoFloat> {
    x0: T,
    y0: T,
    factor: T,
}
impl<T: GeoFloat> DatumCompenseParms<T> {
    pub fn new(hb: T, radius: T, x0: T, y0: T) -> Self {
        let ratio = hb / radius;
        let factor = ratio / (T::ONE + ratio);
        Self { x0, y0, factor }
    }
}
impl<T: GeoFloat> IDatumCompenseParms<T> for DatumCompenseParms<T> {
    fn x0(&self) -> T { self.x0 }
    fn y0(&self) -> T { self.y0 }
    fn factor(&self) -> T { self.factor }
}
/// Converts projected XY coordinates from the height compensation plane to the
/// sea level plane.
///
/// # Arguments
///
/// - `xc`, `yc`: Coordinates on the height compensation plane (in meters).
/// - `hb`: Elevation of the height compensation plane (in meters).
/// - `radius`: Radius of the Earth (in meters).
/// - `x0`, `y0`: Coordinate system origin (in meters).
///
/// # Returns
///
/// A tuple containing the projected XY coordinates of the sea level plane (in
/// meters).
///
///
/// # References
/// - 杨元兴. (2008). 抵偿高程面的选择与计算. 城市勘测 (02), 72-74.
///
/// # Examples
/// ```
/// use float_cmp::assert_approx_eq;
/// let p =(469704.6693, 2821940.796);
/// let parms=pyxis::DatumCompenseParms::new(400.0, 6_378_137.0, 500_000.0, 0.0);
/// let p = pyxis::datum_compense(p.0, p.1, &parms);
/// assert_approx_eq!(f64, p.0, 469706.56912942487, epsilon = 1e-17);
/// assert_approx_eq!(f64, p.1, 2821763.831232311, epsilon = 1e-17);
/// ```
pub fn datum_compense<T>(xc: T, yc: T, parms: &impl IDatumCompenseParms<T>) -> (T, T)
where
    T: GeoFloat,
{
    let xc = xc - parms.factor() * (xc - parms.x0());
    let yc = yc - parms.factor() * (yc - parms.y0());
    (xc, yc)
}
